utils/wasm-gen/src/generator/memory.rs

```diff
@@ -0,0 +1,177 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Memory import generator module.
+
+use crate::{
+    generator::{FrozenGearWasmGenerator, GearWasmGenerator},
+    MemoryPagesConfig, WasmModule,
+};
+use gear_wasm_instrument::{
+    parity_wasm::{
+        builder,
+        elements::{Instruction, Section},
+    },
+    STACK_END_EXPORT_NAME,
+};
+
+/// Memory import generator.
+///
+/// The generator is used to insert into wasm module new
+/// valid (from gear runtime perspective) memory import definition
+/// from the provided config.
+pub struct MemoryGenerator {
+    config: MemoryPagesConfig,
+    module: WasmModule,
+}
+
+impl<'a> From<GearWasmGenerator<'a>> for (MemoryGenerator, FrozenGearWasmGenerator<'a>) {
```

Comment:
I think better to use into_parts in that case

Reply:
shall I implement that in the next PR with `EntryPointsGenerator` and `SysCalls*` generator? In this case I will introduce `into_parts<T>` method, because `GearWasmGenerator` is converted to several generators with a frozen state.